sip-ri/src/main/java/gov/nist/javax/sip/header/extensions/TargetDialog.java

```diff
@@ -0,0 +1,201 @@
+  }
+
+  /**
+   * Sets the local tag of the Target-Dialog header.
+   *
+   * @param localTag the local tag value to set
+   * @throws ParseException if the provided localTag is invalid
+   */
+  public void setLocalTag(String localTag) throws ParseException {
+    if (localTag == null || localTag.trim().isEmpty()) {
+      throw new ParseException("Local tag cannot be null or empty", 0);
+    }
+    this.localTag = localTag.trim();
+  }
+
+  /**
+   * Gets the remote tag of the Target-Dialog header.
+   *
+   * @return the remote tag value
+   */
+  public String getRemoteTag() {
+    return remoteTag;
+  }
+
+  /**
+   * Sets the remote tag of the Target-Dialog header.
+   *
+   * @param remoteTag the remote tag value to set
+   * @throws ParseException if the provided remoteTag is invalid
+   */
+  public void setRemoteTag(String remoteTag) throws ParseException {
+    if (remoteTag == null || remoteTag.trim().isEmpty()) {
+      throw new ParseException("Remote tag cannot be null or empty", 0);
+    }
+    this.remoteTag = remoteTag.trim();
+  }
+
+  @Override
+  public StringBuilder encodeBody(StringBuilder retval) {
+    if (callId != null) {
+      retval.append(callId.encode());
+      if (localTag != null) {
+        retval.append(";"); // Use semicolon directly
+        retval.append(localTag);
+      }
+      if (remoteTag != null) {
+        retval.append(";"); // Use semicolon directly
+        retval.append(remoteTag);
+      }
+      // Encode parameters (if any)
+      if (!parameters.isEmpty()) {
+        retval.append(SEMICOLON).append(parameters.encode());
+      }
+    }
+    return retval;
+  }
+
+  @Override
+  public Object clone() {
+    TargetDialog retval = (TargetDialog) super.clone();
+    if (callId != null) {
+      retval.callId = (CallIdentifier) callId.clone();
+    }
+    retval.localTag = localTag;
+    retval.remoteTag = remoteTag;
+    return retval;
+  }
+
+  @Override
+  public void setValue(String value) throws ParseException {
+    // Not implemented yet, throw an exception to indicate it
+    throw new UnsupportedOperationException("setValue method is not implemented yet");
+  }
+
+  @Override
+  public String getParameter(String name) {
+    return parameters.getValue(name, true).toString();
+  }
+
+  @Override
+  public void setParameter(String name, String value) throws ParseException {
+    parameters.set(name, value);
+  }
+
+  @Override
+  public Iterator<String> getParameterNames() {
+    return parameters.getNames();
+  }
+
+  @Override
+  public void removeParameter(String name) {
+    parameters.delete(name);
+  }
+
+  /**
+   * Parses the header string into Call-Id, local tag, remote tag, and parameters.
+   *
+   * @param body the header string to parse
+   * @throws ParseException if the header string cannot be parsed
+   */
+  public void decodeBody(String body) throws ParseException {
+	    int localTagIndex = body.indexOf(';');
+	    int remoteTagIndex = body.indexOf(';');
```

Comment:
wouldn't both value be the same because body.indexOf would provide the same value ? and thus we would end up with localTag and remoteTag be the same value?

Reply:
Consider this body string: callId;localTag;remoteTag

The first indexOf finds the semicolon after callId, storing 7 in localTagIndex.The second indexOf starts its search from index 8 (after the first semicolon), finding the semicolon after localTag and storing 16 in remoteTagIndex. This approach correctly distinguishes between multiple semicolons and prevents assigning the same value to both localTag and remoteTag. That's why I did this method to avoid ParseExceptions

Reply:
I refactored a lot, I saw what my main problem was, so now it works differently